File: examples/cpa_normal.rs
use anyhow::{Context, Result};
use gnuplot::{Figure, PlotOption::Caption};
use muscat::{distinguishers::cpa_normal::CpaProcessor, leakage_model::aes::sbox};
use ndarray::{Array2, ArrayView1, Axis};
use ndarray_npy::read_npy;
use std::{env, iter::zip, path::PathBuf};

fn leakage_model(plaintext: ArrayView1<usize>, guess: usize) -> usize {
    sbox((plaintext[0] ^ guess) as u8) as usize
}

fn main() -> Result<()> {
    let traces_dir =
        PathBuf::from(env::var("TRACES_DIR").context("Missing TRACES_DIR environment variable")?);

    let traces: Array2<f64> =
        read_npy(traces_dir.join("traces.npy")).context("Failed to read traces.npy")?;
    let plaintexts: Array2<u8> =
        read_npy(traces_dir.join("plaintexts.npy")).context("Failed to read plaintexts.npy")?;
    assert_eq!(traces.shape()[0], plaintexts.shape()[0]);

    // Let's recover the first byte of the key
    let mut processor = CpaProcessor::new(traces.shape()[1], 100, 256);
    for (trace_batch, plaintext_batch) in zip(
        traces.axis_chunks_iter(Axis(0), 100),
        plaintexts.axis_chunks_iter(Axis(0), 100),
    ) {
        processor.update(
            // Convert chipwhisperer float to int
            trace_batch.mapv(|x| ((x + 1.) * 1024.) as u16).view(),
            plaintext_batch,
            leakage_model,
        );
    }

    let cpa = processor.finalize();

    let best_guess = cpa.best_guess();
    println!("Best subkey guess: {:?}", best_guess);

    // Let's plot correlation coefficients of the best guess
    let corr = cpa.corr();
    let corr_best_guess = corr.row(best_guess);

    let mut fg = Figure::new();
    fg.axes2d().lines(
        0..corr_best_guess.len(),
        corr_best_guess,
        &[Caption("Pearson correlation coefficient")],
    );
    fg.show()?;

    Ok(())
}

File: muscatpy/examples/cpa_normal.py
import functools
import os
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

import muscatpy


@functools.cache
def hamming_weight(x: int) -> int:
    weight = 0
    while x != 0:
        if x & 1 == 1:
            weight += 1
        x = x >> 1
    return weight


TRACES_DIR = Path(os.environ["TRACES_DIR"])

traces = np.load(TRACES_DIR / "traces.npy")
plaintexts = np.load(TRACES_DIR / "plaintexts.npy")

# We convert from np.float64 to np.float32 as f64 is not supported
traces = traces.astype(np.float32)
plaintexts = plaintexts.astype(np.uint64)

# Compute the CPA distinguisher using the `compute_cpa_normal` helper
cpa = muscatpy.distinguishers.compute_cpa_normal(
    traces,
    plaintexts,
    256,
    0,
    lambda plaintext, guess: hamming_weight(
        muscatpy.leakage_model.aes.sbox(plaintext ^ guess)
    ),
    200,
)

# Or using the lower level `CpaNormalProcessor` class
leakage_model = lambda plaintext, guess: hamming_weight(
    muscatpy.leakage_model.aes.sbox(int(plaintext[0]) ^ guess)
)
cpa_processor = muscatpy.distinguishers.CpaNormalProcessor(traces.shape[1], 200, 256, traces.dtype)
for i in range(0, traces.shape[0], 200):
    cpa_processor.batch_update(
        traces[i : i + 200], plaintexts[i : i + 200], leakage_model
    )
cpa = cpa_processor.finalize()

best_guess = cpa.best_guess()
print("Best subkey guess:", best_guess)

corr = cpa.corr()[best_guess, :]
plt.plot(corr)
plt.title("Pearson correlation coefficient")
plt.show()
